### ISSUE TITLE
Proposal: Allow the use of preprocessor directives in XAML

### ISSUE THREAD

**Issue by mrlacey** [microsoft/microsoft-ui-xaml#6569]
# Proposal: Allow the use of compiler directives in XAML

## Summary

Compiler preprocessor directives are a powerful tool when working with many programming languages, but they are not available for use in XAML files. Adding them could greatly increase the flexibility of what is created in XAML, reduce code duplication, and allow logic to be better contained in a single XAML file rather than require logic to be split between the XAML file and its accompanying code-behind file.

## Rationale

* Enable the handling of more complex scenarios in XAML files.
* Reduce the need for logic to be split over a `.xaml` file and its accompanying code-behind file (e.g. `.xaml.cs`) 
* Move logic that can be done at compile-time to that point in the process, rather than rely on doing it at run-time so the code is smaller, faster, and less susceptible to error.

## Scope

| Capability | Priority |
| :---------- | :------- |
| This proposal will allow developers to consolidate logic and write less code | Must |

## Important Notes

Such as can be done in other languages, instruction is provided in the code to indicate which parts of the file to include during compilation.

This proposal is to directly mirror the [options available in C#](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives#conditional-compilation) but contain them inside XML comment blocks.

Example:

```xml
<Grid>
<!-- #if $(DEBUG) -->
    <TextBlock Text="In debug mode" />
<!-- #elif $(RELEASE) -->
    <TextBlock Text="In release mode" />
<!-- #else -->
    <TextBlock Text="In some other mode" />
<!-- #endif -->
</Grid>
````

Assuming the above was in the user authored file, and then the code was compiled with the `DEBUG` directive specified, the version of the XAML that was passed to the compiler (language service) and include in the assembly would be:

```xml
<Grid>
    <TextBlock Text="In debug mode" />
</Grid>
```

The comments and anything not appropriate to include based on the available directives should be invisible to the rest of the compilation process and compiled application.

---

I suggest putting the directives inside comment blocks, so they are ignored by any other tooling that looks at XAML files but doesn't know about this capability. Adding something entirely new could risk breaking lots of other things...

Individual directives are indicated by specifying them between `$(` and `)` to reflect the notation used in msbuild files. 

---

The definition (or setting) of directives within a XAML file must also be supported.  This should be in a similar manner to that explained above.

example:

```xml
<!-- #define $(DEMO) -->
...
<!-- #if $(DEMO) -->
    <TextBlock Text="In demo mode" />
<!-- #endif -->
```

The use of `$(` and `)` in the definition is to avoid any possible confusion about the inclusion of whitespace and the closing comment characters (`-->`).

One or more whitespace characters should come after `<!--` characters and before the directive.
One or more whitespace characters should come after the directive and before the closing `-->` characters.
Directives in comments should always be in a single line and not be supported in comment blocks that go over multiple lines.

Directives specified at the project level (inside `.*proj` files) should be available within all the XAML files inside that project. It should not be necessary to re-specify these in the XAML field where used.

## Open Questions

 
* Is it necessary to support logic in the evaluation of an `#if` or `#elif` block to check for more than one directive at a time? (If it is this would presumably be done by using `|` between directives to indicate **OR** and `&` between directives to indicate **AND**. If supported how advanced should this logic be and would it need additional brackets to manage precedence?)
* Is it necessary to be able to unset a directive in XAML as well as set it? (I don't see a need but there may be one. It might also be good to include this for completeness.)
* Is it necessary or appropriate to require at least one whitespace character after `<!--` and before `-->`?
* If enabling the functionality in this proposal made it possible to include other functionality in a XAML pre-processor step, could this also be made extensible? (i.e. Could I write my own tasks or extensions that look at and modify the XAML file before the built-in processor (language service) looks at it?)

* Is this the best place for this proposal? - I'll happily ask it elsewhere as I recognize that it relates to functionality that is not and will not be open-sourced and made fully public.

**Comment by sonnemaf** [comment]
This would indeed be very nice to have. Not sure if it is a MUST HAVE though.

**Comment by mrlacey** [comment]
> Not sure if it is a MUST HAVE though.

Do you mean in relation to other proposals, or the classification of priority for the capability within the scope?

**Comment by sonnemaf** [comment]
Yes I mean in relation with other proposals. There are so many open Issues and items on the roadmap that have an higher priority for ME. I'm glad that I don't have to decide which one comes first.

**Comment by knat** [comment]
Code is the king, XAML is not(it's markup). Follow Charles Petzold, creating views in pure C#/C++ code😄.

**Comment by mrlacey** [comment]
Also raised as a discussion point for inclusion in .NET MAUI https://github.com/dotnet/maui/discussions/4096

**Comment by michael-hawker** [comment]
I wonder if it'd be easier to add-on/follow the conditional XAML approach that was taken for windows versioning?

e.g.

```xml
<Page xmlns:debug="http://schemas.microsoft.com/winfx/2006/xaml/presentation?IsDirectivePresent(DEBUG)"
           xmlns:release="http://schemas.microsoft.com/winfx/2006/xaml/presentation?IsDirectivePresent(RELEASE)">
  <StackPanel>
     <debug:TextBlock Text="This only shows up in Debug Mode"/>
     <TextBlock Text="This will be bold in Debug Mode" debug:FontWeight="Bold"/>
     <release:Image Source="..."/> <!-- This image only shows up in release mode -->
```

This would feel familiar to existing patterns for developers and also allow a bit more fine-tuned control over specific properties vs. large chunks with regions. You also get immediate context with each element when it is used with which directive it will be attached to vs. having to know it's in a special comment block.

**Comment by mrlacey** [comment]
> I wonder if it'd be easier to add-on/follow the conditional XAML approach that was taken for windows versioning?

While this is nice, it makes the check a runtime one, rather than a build one, and so sidesteps my ulterior motive of eventually opening up the compilation process to extensibility 😉

**Comment by michael-hawker** [comment]
@mrlacey I don't think it would have to be a runtime check. The XAML Compiler already is pulling out `{x:Bind}` to hook in compile time code with the XAML syntax, so I don't see how it looking for this special function would be different. I can't imagine the [Design Time Data](https://docs.microsoft.com/en-us/visualstudio/xaml-tools/xaml-designtime-data) feature which does similar xmlns shenanigans is a runtime check, eh?

Though I don't know how any of that works, so I could be wrong here...

**Comment by mrlacey** [comment]
> @mrlacey I don't think it would have to be a runtime check. The XAML Compiler already is pulling out `{x:Bind}` to hook in compile time code with the XAML syntax, so I don't see how it looking for this special function would be different. I can't imagine the [Design Time Data](https://docs.microsoft.com/en-us/visualstudio/xaml-tools/xaml-designtime-data) feature which does similar xmlns shenanigans is a runtime check, eh?
> 
> Though I don't know how any of that works, so I could be wrong here...

Compiler directives are not normally available at runtime so this would be a change to the way compiler directives work and that may be confusing. The current implementation of conditional XAML is that it is resolved at runtime (so it can do things like identifying which OS version it is running on.)

I believe that design-time data is treated internally as a special case, but again that's not being evaluated at runtime.

Of course, [if they gave us a way to run our own code in conditional XAML functions](https://github.com/microsoft/microsoft-ui-xaml/issues/6546#issuecomment-1009022230) then we could build our own workaround solution. 😉

**Comment by StephenLPeters** [comment]
The conditional xaml in winui2 is a custom implementation which is handled by preprocessing the xaml file and generating different xaml files for the different contracts.  It is indeed very usful.  making this standard sounds cool.  @licanhua and @fabiant3 FYI
